packages/server/graphql/queries/helpers/fetchAtlassianProjects.ts

```diff
@@ -1,46 +1,40 @@
-import {DataLoaderWorker} from '../../graphql'
+import {GQLContext} from './../../graphql'
 import AtlassianServerManager from '../../../utils/AtlassianServerManager'
-import makeJiraProjectName from 'parabol-client/utils/makeJiraProjectName'
-import makeSuggestedIntegrationId from 'parabol-client/utils/makeSuggestedIntegrationId'
-import {TaskServiceEnum} from '../../../database/types/Task'
+import {getUserId} from '../../../utils/authorization'
+import IntegrationRepoId from 'parabol-client/shared/gqlIds/IntegrationRepoId'
 
-const fetchAtlassianProjects = async (
-  dataLoader: DataLoaderWorker,
-  teamId: string,
-  userId: string
-) => {
-  const auth = await dataLoader.get('freshAtlassianAuth').load({teamId, userId})
-  if (!auth) return []
-  const {accessToken} = auth
-  const manager = new AtlassianServerManager(accessToken)
+const getCloudIds = async (manager: AtlassianServerManager) => {
   const sites = await manager.getAccessibleResources()
-
   if ('message' in sites) {
     console.error(sites)
-    return []
+    return null
   }
+  return sites.map(({id}) => id)
+}
 
-  const cloudIds = sites.map((site) => site.id)
-  const atlassianProjects = [] as any
-  const service: TaskServiceEnum = 'jira'
-  await manager.getProjects(cloudIds, (err, res) => {
-    if (err) {
-      console.error(err)
-    } else if (res) {
-      const {cloudId, newProjects} = res
-      const newItems = newProjects.map((project) => ({
-        // projectId/key is not globally unique, but a cloudId is
-        id: makeSuggestedIntegrationId({...project, projectKey: project.key, cloudId, service}),
-        service,
-        cloudId,
-        projectName: makeJiraProjectName(project.name, sites, cloudId),
-        projectKey: project.key,
-        avatar: project.avatarUrls['24x24']
-      }))
-      atlassianProjects.push(...newItems)
-    }
-  })
-  return atlassianProjects
+const fetchAtlassianProjects = async (
+  teamId: string,
+  userId: string,
+  {dataLoader, authToken}: GQLContext,
+  cloudIds?: string[],
+  accessToken?: string
+) => {
+  const viewerId = getUserId(authToken)
+  if (viewerId !== userId) return []
+  const token =
+    accessToken || (await dataLoader.get('freshAtlassianAuth').load({teamId, userId}))?.accessToken
+  if (!token) return []
+  const manager = new AtlassianServerManager(token)
+  const jiraCloudIds = cloudIds || (await getCloudIds(manager))
+  if (!jiraCloudIds) return []
+  const projects = await manager.getAllProjects(jiraCloudIds)
+  return projects.map((project) => ({
+    ...project,
+    id: IntegrationRepoId.join({...project, projectKey: project.key, service: 'jira'}),
```

Comment:
+1 passing in `cloudId` instead of a rest param is a little more expressive because we don't have to dive into the child function to know what fields it actually needs.

+1 since the service is already known, you can use `JiraProjectId.join` instead.

Reply:
+1 what are your thoughts on turning this whole function into a dataloader?
all it would need is `{teamId, userId}`. it could call the `freshAtlassianAuth` dataloader to get the auth, and then `atlassianCloudNameLookup` dataloader to get the cloudIds, and then call `getAllProjects`. that could eliminate a roundtrip to atlassian for the cloudIds in some cases & we could get rid of `getCloudIds`

Reply:
I like it! I've refactored this to a dataloader. We previously discussed with Georg that it's best not to return the inputs unchanged in the outputs of a dataloader. So, I'm just returning `manager.getAllProject` in the dataloader without the `userId` or `teamId`. I think that's OK, but it does concern me that [JiraRemoteProject](https://github.com/ParabolInc/parabol/blob/0c0a286dd28266dc1eec710029af6780574b1a3b/packages/server/graphql/types/JiraRemoteProject.ts#L29) requires a `userId` and `teamId` so we have to always remember modify the returned values of the dataloader.

Reply:
Up to you! As a challenge question: What's bad about returning an input value in the output? Every `tasksByTeamId` accepts a teamId & returns a task with the teamId on it. 

- teamId, userId are required inputs whether or not they're included in the output
- Every call to the dataloader eventually returns a JiraRemoteProject which requires teamId, userId
- we already do it with cloudId & `manager.getAllProjects`

it's less code, no chance of forgetting to add a mapper.

Reply:
I'd say that returning an input value unchanged as the output is arguably confusing as we can just include it in the place where we're calling the loader. But, I'll go with my original instinct of including it in the dataLoader because I think we could run into bugs where `RemoteJiraProject` doesn't include `userId` or `teamId`, and, as you say, it's less code.